classes/core/PKPApplication.inc.php

```diff
@@ -292,35 +292,35 @@ function getName() {
 	 * @return array
 	 */
 	function &getEnabledProducts($category = null, $mainContextId = null) {
-		if (is_null($this->enabledProducts) || !is_null($mainContextId)) {
-			$contextDepth = $this->getContextDepth();
-
+		$contextDepth = $this->getContextDepth();
+		if (is_null($mainContextId)) {
+			$request = $this->getRequest();
+			$router = $request->getRouter();
+
+			if (is_null($mainContextId)) {
+				// Try to identify the main context (e.g. journal, conference, press),
+				// will be null if none found.
+				$mainContext = $router->getContext($request, 1);
+				if ($mainContext) $mainContextId = $mainContext->getId();
+			}
+		}
+		if (!isset($this->enabledProducts[$mainContextId])) {
 			$settingContext = array();
 			if ($contextDepth > 0) {
-				$request = $this->getRequest();
-				$router = $request->getRouter();
-
-				if (is_null($mainContextId)) {
-					// Try to identify the main context (e.g. journal, conference, press),
-					// will be null if none found.
-					$mainContext = $router->getContext($request, 1);
-					if ($mainContext) $mainContextId = $mainContext->getId();
-				}
-
 				// Create the context for the setting if found
-				if (!is_null($mainContextId)) $settingContext[] = $mainContextId;
+				$settingContext[] = $mainContextId;
 				$settingContext = array_pad($settingContext, $contextDepth, 0);
 				$settingContext = array_combine($this->getContextList(), $settingContext);
 			}
 
 			$versionDao = DAORegistry::getDAO('VersionDAO'); /* @var $versionDao VersionDAO */
-			$this->enabledProducts = $versionDao->getCurrentProducts($settingContext);
+			$this->enabledProducts[$mainContextId] = $versionDao->getCurrentProducts($settingContext);
```

Comment:
What if $mainContextId is null? It will be converted to "" array key and will then be overridden every time the $mainContextId is null, if it is cached?, which is maybe OK and works?
In this case, the products should be from journal = 0 (see the comment above)? (That was also earlier the case.) Or this case when $mainContext does not exist (which is surely almost never and maybe an error case, but still possible?) should be handled somehow differently?
If in this case products should be from journal = 0, maybe it would be better to have this $mainContextId array key = 0 in this case $mainContextId is null?

Reply:
As above, I'll set it to CONTEXT_SITE (0) in that case.